src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -331,98 +356,107 @@ public Assembly LoadFromPath(string originalAnalyzerPath)
-        /// </summary>
-        internal string GetRealAnalyzerLoadPath(string originalFullPath)
+        internal (string OriginalAssemblyPath, string ResolvedAssemblyPath)[] GetPathMapSnapshot()
         {
             CheckIfDisposed();
 
             lock (_guard)
             {
-                if (!_analyzerAssemblyInfoMap.TryGetValue(originalFullPath, out var tuple))
-                {
-                    throw new InvalidOperationException($"Invalid original path: {originalFullPath}");
-                }
-
-                return tuple is { } value ? value.RealAssemblyPath : originalFullPath;
+                return _resolvedToOriginalPathMap.Select(x => (x.Value, x.Key)).ToArray();
             }
         }
 
-        internal (string OriginalAssemblyPath, string RealAssemblyPath)[] GetPathMapSnapshot()
+#if NET
+        /// <summary>
+        /// Return an <see cref="IAnalyzerAssemblyLoader"/> which does not lock assemblies on disk that is
+        /// most appropriate for the current platform.
+        /// </summary>
+        /// <param name="windowsShadowPath">A shadow copy path will be created on Windows and this value 
+        /// will be the base directory where shadow copy assemblies are stored. </param>
+        internal static IAnalyzerAssemblyLoaderInternal CreateNonLockingLoader(
+            string windowsShadowPath,
+            ImmutableArray<IAnalyzerPathResolver> pathResolvers = default,
+            ImmutableArray<IAnalyzerAssemblyResolver> assemblyResolvers = default,
+            System.Runtime.Loader.AssemblyLoadContext? compilerLoadContext = null)
         {
-            CheckIfDisposed();
+            pathResolvers = pathResolvers.NullToEmpty();
+            assemblyResolvers = assemblyResolvers.NullToEmpty();
 
-            lock (_guard)
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                // The only reason we don't use LoadFromStream on Windows is because the anti-virus checker will amortize
+                // the cost of scanning files on disk. When loading from stream there is no such amortization, the 
+                // anti-virus checker will scan the file every time it is loaded and the performance hit is significant.
+                //
+                // On non-Windows OS this is generally not a concern and loading from stream is sufficient to avoid
+                // locking the file.
+                return new AnalyzerAssemblyLoader(
+                    pathResolvers,
+                    [.. assemblyResolvers, StreamResolver.Instance],
+                    compilerLoadContext);
+            }
+
+            // The shadow copy analyzer should only be created on Windows. To create on Linux we cannot use 
+            // GetTempPath as it's not per-user. Generally there is no need as LoadFromStream achieves the same
+            // effect
+            if (!Path.IsPathRooted(windowsShadowPath))
```

Comment:
Would it make sense to move this check into the constructor of ShadowCopyAnalyzerPathResolver?

Reply:
Yeah that is a much better place